llvm/lib/Analysis/ScalarEvolution.cpp

```diff
@@ -10857,7 +10861,11 @@ bool ScalarEvolution::isKnownToBeAPowerOfTwo(const SCEV *S, bool OrZero) {
   auto *Mul = dyn_cast<SCEVMulExpr>(S);
   if (!Mul)
     return false;
-  return all_of(Mul->operands(), NonRecursive) && (OrZero || isKnownNonZero(S));
+  return all_of(Mul->operands(), NonRecursive) &&
+         (OrZero || isKnownNonZero(S)) &&
+         (!OrNegative || llvm::count_if(Mul->operands(), [this](const SCEV *S) {
```

Comment:
I don't get why we need this condition.

If OrNegative=false the previous code is correct. If OrNegative=true, then we're fine with either a positive or negative power of two. So even if we have two negative factors, the result will be a positive power of two (or zero) and still be fine.

Reply:
You're right.  I was originally thinking of this as power-of-two XOR negative power-of-two, and then generalized slightly, but you're right, this condition is a remnant of the XOR model.  Will kill shortly.